enhancements/authentication/oauth-resource-storage.md

```diff
@@ -0,0 +1,197 @@
+and hence making the keys alone insensitive.
+
+The token passed to the clients will be the concatenated object name plus the new
+secret string in the body, i.e. a compound token.
+
+The authenticator in kube-apiserver will be extended to make the lookup to the
+oauth access tokens with the first half of the passed token from the client, and
+then compares the body as well before authenticating the request.
+
+## Motivation
+
+Etcd encryption without encrypted oauth access and authorize token is of limited
+value. While these tokens expire after a day usually, the token from a fresh backup
+or the output of etcdctl can be used to act as any oauth authenticated user in the 
+cluster.
+
+A side-effect of the proposed compound tokens is that it allows to hide the secret part
+in the body on LIST or GET requests without returning sensitive information. This
+way users could be allowed to read and possible delete their oauth access tokens, 
+e.g. in order to show that information via oc or console, or to logout certain tokens.
+
+### Goals
+
+- Change the `oauthaccesstokens` and `oauthauthorizetokens` storage format and schema such that
+  the etcd keys alone (which are not encrypted when encryption at REST is enabled) are not sensitive.
+- Allow console and oc to log out all other sessions.
+
+### Non-Goals
+
+## Proposal
+
+We extend the OAuthAccessToken and OAuthAuthorizeToken types in the following way:
+
+```go
+// OAuthAccessToken describes an OAuth access token
+type OAuthAccessToken struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    ...
+    
+    // SecretToken is the second half of the secret information a user has to pass as bearer token
+    // in order to authorize with the API server. It is concatenated with the name of the 
+    // OAuthAccessToken object.
+    // For old tokens, this field might be empty.
+    // +optional
+    SecretToken string `json:"secretToken,omitempty" protobuf:"bytes,11,opt,name=secretToken"`
+}
+
+// OAuthAuthorizeToken describes an OAuth authorization token
+type OAuthAuthorizeToken struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    ...
+    
+    // SecretToken is the second half of the secret information a user has to pass as bearer token
+    // in order to authorize with the API server. It is concatenated with the name of the 
+    // OAuthAuthorizeToken object.
+    // For old tokens, this field might be empty.
+    // +optional
+    SecretToken string `json:"secretToken,omitempty" protobuf:"bytes,11,opt,name=secretToken"`
+}
+```
+
+### User Stories [optional]
+
+#### Story 1
+
+As an attacker who got access to an etcd backup I do not get access to oauth tokens.
+
+#### Story 2
+
+As a user I can use the API via oc or console to log out other sessions. 
```

Comment:
all other sessions?

there's a difference. You need to be sure that selective removal isn't required (I bet it is...)

Reply:
Yep. But this story is about the simple case of logging everything out other than the current token.

Reply:
Referring this CWE - https://cwe.mitre.org/data/definitions/613.html we can start this by removing all other sessions. Haven't heard any customer request for selective removal. Would keep things simple and provide option for remove all

Reply:
Made out-of-scope of this PR and move to another release.